File: example/cxx/dense/frank_cartesian_comm_mpi.cpp
/*****************************************************************************
*
* Rokko: Integrated Interface for libraries of eigenvalue decomposition
*
* Copyright (C) 2012-2016 Rokko Developers https://github.com/t-sakashita/rokko
*
* Distributed under the Boost Software License, Version 1.0. (See accompanying
* file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
*
*****************************************************************************/

#include <rokko/rokko.hpp>
#include <rokko/mapping_bc.hpp>
#include <rokko/collective.hpp>
#include <rokko/utility/solver_name.hpp>
#include <rokko/utility/frank_matrix.hpp>
#include <boost/lexical_cast.hpp>
#include <iostream>


typedef rokko::matrix_col_major matrix_major;

int main(int argc, char *argv[]) {
  int provided;
  MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
  
  MPI_Comm comm;
  int dims[2], period[2];
  dims[0] = 2;  dims[1] = 2;
  period[0] = 0;  period[1] = 0;
  int reorder = 0;

  int ierr = MPI_Cart_create(MPI_COMM_WORLD, 2, dims, period, reorder, &comm);
  rokko::grid g(comm);
  int myrank = g.get_myrank();

  if (comm != MPI_COMM_NULL) {
    std::string library_routine(rokko::parallel_dense_ev::default_solver());
    std::string library, routine;
    int dim = 10;
    if (argc >= 2) library_routine = argv[1];
    if (argc >= 3) dim = boost::lexical_cast<int>(argv[2]);
    rokko::split_solver_name(library_routine, library, routine);
    
    std::cout.precision(5);
    
    rokko::parallel_dense_ev solver(library);
    solver.initialize(argc, argv);

    if (myrank == 0)
      std::cout << "Eigenvalue decomposition of Frank matrix" << std::endl
		<< "library:routine = " << library_routine << std::endl
		<< "num_procs = " << g.get_nprocs() << std::endl
                #ifdef _OPENMP
		<< "num_threads per process = " << omp_get_max_threads() << std::endl
                #endif
		<< "routine = " << routine << std::endl
		<< "dimension = " << dim << std::endl;
    
    rokko::mapping_bc<matrix_major> map = solver.default_mapping(dim, g);
    rokko::distributed_matrix<double, matrix_major> mat(map);
    rokko::frank_matrix::generate(mat);
    rokko::localized_matrix<double, matrix_major> mat_loc(dim, dim);
    rokko::gather(mat, mat_loc, 0);
    
    rokko::localized_vector<double> eigval(dim);
    rokko::distributed_matrix<double, matrix_major> eigvec(map);
    rokko::parameters params;
    params.set("routine", routine);
    try {
      solver.diagonalize(mat, eigval, eigvec, params);
    } catch (const char *e) {
      if (myrank == 0) std::cout << "Exception : " << e << std::endl;
      MPI_Abort(MPI_COMM_WORLD, 22);
    }
    
    rokko::localized_matrix<double, matrix_major> eigvec_loc(dim, dim);
    rokko::gather(eigvec, eigvec_loc, 0);
    if (myrank == 0) {
      bool sorted = true;
      for (unsigned int i = 1; i < dim; ++i) sorted &= (eigval(i-1) <= eigval(i));
      if (!sorted) std::cout << "Warning: eigenvalues are not sorted in ascending order!\n";
      
      std::cout << "largest eigenvalues:";
      for (int i = 0; i < std::min(dim, 10); ++i) std::cout << ' ' << eigval(dim - 1 - i);
      std::cout << std::endl;
      std::cout << "residual of the largest eigenvalue/vector: |x A x - lambda| = "
		<< std::abs(eigvec_loc.col(dim - 1).transpose() * mat_loc * eigvec_loc.col(dim - 1)
			    - eigval(dim - 1))
		<< std::endl;
    }    
    solver.finalize();
    MPI_Comm_free(&comm);
  }
  MPI_Finalize();
}

File: example/fortran/dense/frank_cartesian_comm_mpi.f90
!*****************************************************************************
!
! Rokko: Integrated Interface for libraries of eigenvalue decomposition
!
! Copyright (C) 2012-2016 by Rokko Developers https://github.com/t-sakashita/rokko
!
! Distributed under the Boost Software License, Version 1.0. (See accompanying
! file LICENSE_1_0.txt or copy at http://www.boost.org/license_1_0.txt)
!
!*****************************************************************************

program frank_matrix
  use rokko
  implicit none
  include 'mpif.h'
  integer :: dim
  type(rokko_distributed_matrix) :: mat, Z
  type(rokko_grid) :: grid
  type(rokko_mapping_bc) :: map
  type(rokko_parallel_dense_ev) :: solver
  type(rokko_eigen_vector) :: w
  character(len=20) :: library, routine
  character(len=100) :: library_routine, tmp_str
  integer arg_len, status

  integer :: provided, ierr, myrank, nprocs
  integer :: i
  integer :: comm
  integer, dimension(2) :: dims
  logical, dimension(2) :: periods 
  logical :: reorder
  
  call MPI_init_thread(MPI_THREAD_MULTIPLE, provided, ierr)
  call MPI_comm_rank(MPI_COMM_WORLD, myrank, ierr)
  call MPI_comm_size(MPI_COMM_WORLD, nprocs, ierr)
  dims(1) = int(sqrt(real(nprocs)));
  do while (.true.)
     if ( dims(1) == 1 ) exit
     if ( mod(nprocs, dims(1)) == 0 ) exit
     dims(1) = dims(1) - 1
  enddo
  dims(2) = nprocs / dims(1);
  periods(1) = .false.;  periods(2) = .false.
  reorder = .false.
  call mpi_cart_create(MPI_COMM_WORLD, 2, dims, periods, reorder, comm, ierr)
  if (myrank == 0) then
     write(*,'("Created ", i0, "x", i0, " size communicator with new cartesian topology")') dims(1), dims(2)
  endif

  if (comm /= MPI_COMM_NULL) then
     if (command_argument_count() >= 1) then
        call get_command_argument(1, library_routine, arg_len, status)
     else
        call rokko_parallel_dense_ev_default_solver(library_routine)
     endif
     call rokko_split_solver_name(library_routine, library, routine)
     
     if (command_argument_count() == 2) then  
        call get_command_argument(2, tmp_str, arg_len, status)
        read(tmp_str, *) dim
     else
        dim = 10
     endif

     if (myrank == 0) then
        print *,"library = ", library
        print *,"routine = ", routine
        print *,"dimension = ", dim
     endif
     
     call rokko_construct(solver, library)
     call rokko_construct(grid, comm, rokko_grid_row_major)
     call rokko_default_mapping(solver, dim, grid, map)
     call rokko_construct(mat, map)
     call rokko_construct(Z, map)
     call rokko_construct(w, dim)
     
     ! generate frank matrix
     call rokko_frank_matrix_generate(mat)
     call rokko_print(mat)
     
     call rokko_diagonalize(solver, mat, w, Z)
     
     if (myrank.eq.0) then
        write(*,'(A)') "Computed Eigenvalues = "
        do i = 1, dim
           write(*,"(f30.20)") rokko_eigen_vector_get(w, i)
        enddo
     endif
     
     call rokko_destruct(mat)
     call rokko_destruct(Z)
     call rokko_destruct(w)
     call rokko_destruct(solver)
     call rokko_destruct(map)
     call rokko_destruct(grid)
     
     call mpi_comm_free(comm, ierr)
  endif
  call MPI_finalize(ierr)
end program frank_matrix
